### ISSUE TITLE
add react-dom as a peer dependency (yarn 2 compatibility)

### ISSUE THREAD

**Issue by PutziSan** [pmndrs/jotai#135]
When using this package with yarn v2 (yarn berry), it throws while building:

```
Module not found: Error: use-context-selector tried to access react-dom (a peer dependency) but it isn't provided by its ancestors; this makes the require call ambiguous and unsound.
```

> this error comes after resolving #134 by adding scheduler as a dependency ;-) (resolving this error is a bit easier than #134 - so I created another issue, hope this is fine for you)

You depend on [use-context-selector](https://www.npmjs.com/package/use-context-selector), and they declared `react-dom` as a peerDependency. So you should also add `react-dom` to your peerDependencies to make sure, it is available.

## workaround at the moment

Currently workaround is adding this to `.yarnrc.yml`

```yaml
packageExtensions:
  jotai@*:
    peerDependencies:
      "react-dom": "*"
```

info about packageExtensions:

> Some packages may have been specified incorrectly with regard to their dependencies - for example with one dependency being missing, causing Yarn to refuse it the access. The packageExtensions fields offer a way to extend the existing package definitions with additional information.

https://yarnpkg.com/configuration/yarnrc#packageExtensions

**Comment by dai-shi** [comment]
Oh, OK. So the issue is not only about scheduler but the peer dependency of use-context-selector.
I'd like to know if this issue is only with yarn 2. We also need to take care of RN users.

Does your app package.json not include react-dom? Does it give you an error without the workaround? 🤔

**Comment by dai-shi** [comment]
> resolving this error is a bit easier than #134

Could you explain why it's easier? Does `packageExtensions` not work for `scheduler`?

**Comment by PutziSan** [comment]
This issue will be resolved when jotai adds `react-dom` as a peerDependency in package.json. I do not see any disadvantage in doing this. 

In #134 you must decide, what you want to do with the scheduler dependency, this decision seems not that trivial to me.

**Comment by dai-shi** [comment]
Thanks. For React Native users, they don't install react-dom. Actually, peerDependencyMeta specifies react-dom as optional.

**Comment by dai-shi** [comment]
So, for now our recommendation is the workaround described in https://github.com/pmndrs/jotai/issues/135#issue-723129121

```yaml
packageExtensions:
  jotai@*:
    peerDependencies:
      "react-dom": "*"
```

or the one described in https://github.com/pmndrs/jotai/pull/139#issuecomment-710701325

```json
  "dependencies": {
    "react": "^16.14.0",
    "react-dom": "^16.14.0"
  },
```

**Comment by merceyz** [comment]
> Thanks. For React Native users, they don't install react-dom. Actually, peerDependencyMeta specifies react-dom as optional.

In that case you need to declare `react-dom` as an optional peer dependency.

> I'd like to know if this issue is only with yarn 2.

This is an issue in general, Yarn 2's PnP mode is just really strict about it. Not declaring it as a peer dependency means that `use-context-selector` relies on hoisting to bring `react-dom` to a level where it can access it, which is not guaranteed to happen and relies on undefined behaviour.

**Comment by dai-shi** [comment]
What I'm not sure is if yarn 2 really requires an _intermediate_ library package to declare all peer dependencies of its children, grand children and grand grand children. It sounds unlikely to me.

**Comment by merceyz** [comment]
It does, please give https://dev.to/arcanis/implicit-transitive-peer-dependencies-ed0 a read on why

**Comment by dai-shi** [comment]
> TL;DR: If you write a package that depends on Foo, and if Foo has a peer dependency, then you must provide it in either of the dependencies or peerDependencies fields. You won't "implicitly inherit" the peer dependencies declared in Foo.

Alright, I wasn't aware this. Thanks for noting.
